### PR TITLE
New commands: `azdev scan/mask` for secrets scanning and redaction

### PR DESCRIPTION
**Pull request by evelyn-ys**
We have more and more scenarios which require a secrets handling tool. So add
- `azdev scan` command for scan
- `azdev mask` command for redaction

### CODE REVIEW

File: azdev/operations/secret.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+
+def _validate_data_path(file_path=None, directory_path=None, data=None):
+    if file_path and directory_path:
+        raise ValueError('Can not specify file path and directory path at the same time')
+    if file_path and data:
+        raise ValueError('Can not specify file path and raw string at the same time')
+    if directory_path and data:
+        raise ValueError('Can not specify directory path and raw string at the same time')
+    if not file_path and not directory_path and not data:
+        raise ValueError('No file path ot directory path or raw string provided')
```
Comment (wangzelin007) [inline azdev/operations/secret.py]:
```suggestion
        raise ValueError('No file path or directory path or raw string provided')
```

File: azdev/operations/secret.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+    return load_regex_patterns_from_json_file('PreciselyClassifiedSecurityKeys.json')
+
+
+def _load_regex_patterns(custom_pattern=None):
+    built_in_regex_patterns = _load_built_in_regex_patterns()
+
+    if not custom_pattern:
+        return built_in_regex_patterns
+    if os.path.isfile(custom_pattern):
+        custom_pattern = json.load(custom_pattern)
```
Comment (wangzelin007) [inline azdev/operations/secret.py]: Does custom_pattern have to be in JSON format? Do we need to provide additional error handling, such as capturing json.JSONDecodeError and providing a more friendly error prompt?

File: azdev/params.py [hunk 3]
```diff
@@ -100,6 +100,32 @@ def load_arguments(self, _):
...
+            c.argument('file_path', options_list=['--file-path', '-f'],
+                       help='Path of the file you want to scan secrets for')
+            c.argument('directory_path', options_list=['--directory-path', '-d'],
+                       help='Path of the folder you want to scan secrets for')
+            c.argument('recursive', options_list=['--recursive', '-r'],
+                       help='Scan the directory recursively')
+            c.argument('data', help='Raw string you want to scan secrets for')
+            c.argument('save_scan_result', options_list=['--save-scan_result', '--save'], type=bool,
+                       help='Whether to save scan result to file or not')
+            c.argument('scan_result_path', options_list=['--scan-result-path', '--result'],
```
Comment (wangzelin007) [inline azdev/params.py:115]: Is it necessary to provide more explanation on the relationship between the two options `--save-scan-result` and `--scan-result-path` to help users understand their purpose and relationship with each other, and inform users of the default path for saving?

File: azdev/help.py [hunk 4]
```diff
@@ -159,6 +160,51 @@
...
+                azdev scan --data "my string waiting to be scanned" --save-scan-result True
+        - name: Recursively scan secrets for a directory and save results to specific file
+          text: |
+                azdev scan --directory-path /path/to/my/folder --recursive --scan-result-path /path/to/scan_result.json
+"""
+
+helps['mask'] = """
+    short-summary: Mask secrets for files or string
+    long-summary: |
+                Redaction type 'FIXED_VALUE' will mask all secrets with '+++'.
```
Comment (wangzelin007) [inline azdev/help.py]: Is it more common to mask secrets with '***'?
What are the considerations behind using the '+++' ?

Comment (evelyn-ys) [reply azdev/help.py]: '+++' is default mask used by 1ES SDK. But it's true that '***' is more common.

File: azdev/operations/secret.py (hunk 2 above)
Comment (evelyn-ys) [reply azdev/operations/secret.py]: added

File: azdev/params.py (hunk 3 above)
Comment (evelyn-ys) [reply azdev/params.py:115]: added detailed help msg

File: azdev/help.py (hunk 4 above)
Comment (evelyn-ys) [reply azdev/help.py]: changed

**Review by wangzelin007 — APPROVED** [review]

**Pull request merged**